src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs

```diff
@@ -220,22 +220,42 @@ internal void Clear()
             lock (_readSyncObject)
             {
                 priorityType = DataPriorityType.Default;
-
-                // send data from which ever stream that has data directly.
                 byte[] result = null;
-                result = _dataToBeSent[(int)DataPriorityType.PromptResponse].ReadOrRegisterCallback(_onSendCollectionDataAvailable);
-                priorityType = DataPriorityType.PromptResponse;
 
-                if (result == null)
-                {
-                    result = _dataToBeSent[(int)DataPriorityType.Default].ReadOrRegisterCallback(_onSendCollectionDataAvailable);
-                    priorityType = DataPriorityType.Default;
-                }
-                // no data to return..so register the callback.
-                if (result == null)
+                if (_dataSyncObjects != null && _dataToBeSent != null)
                 {
-                    // register callback.
-                    _onDataAvailableCallback = callback;
+                    const int promptResponseIndex = (int)DataPriorityType.PromptResponse;
+                    lock (_dataSyncObjects[promptResponseIndex])
```

Comment:
Actually, I was thinking about this some more, and I don't think this is a safe fix.  The `_dataSyncObjects` sync objects are intended to serialize adding data to send, but taking a lock during stream access invites deadlock scenarios.  

This race condition exception occurs on a race between a remote command complete and connection close events.  In this case the data-to-send is likely empty.  But taking a lock may deadlock, so I feel better pattern is to not take the lock but check for null reference.  The SerializedDataStream object already accounts for if it is disposed, so null ref check is all that is needed.  So I am thinking of something like:

```csharp
internal byte[] ReadOrRegisterCallback(
    OnDataAvailableCallback callback,
    out DataPriorityType priorityType)
{
    priorityType = DataPriorityType.Default;

    lock (_readSyncObject)
    {
        // Send data from which ever stream that has data directly.
        byte[] result = null;
        SerializedDataStream promptDataToBeSent = _dataToBeSent[(int)DataPriorityType.PromptResponse];
        if (promptDataToBeSent is not null)
        {
            result = promptDataToBeSent.ReadOrRegisterCallback(_onSendCollectionDataAvailable);
            priorityType = DataPriorityType.PromptResponse;
        }

        if (result == null)
        {
            SerializedDataStream defaultDataToBeSent = _dataToBeSent[(int)DataPriorityType.Default];
            if (defaultDataToBeSent is not null)
            {
                result = defaultDataToBeSent.ReadOrRegisterCallback(_onSendCollectionDataAvailable);
                priorityType = DataPriorityType.Default;
            }
        }

        // No data to return..so register the callback.
        if (result == null)
        {
            // Register callback.
            _onDataAvailableCallback = callback;
        }

        return result;
    }
}
```

@SergeyZalyadeev  can you try this code variation with your tests?  We could also add try/catch for `ObjectDisposed` exceptions, but I am hoping it is not necessary.

/cc: @daxian-dbw

Reply:
@PaulHigin, it works in my labs.

Reply:
I agree that we should try to avoid applying new locks in this method as much as possible. I think doing null-checking on `_dataToBeSent[<index>]` should take care of the `_dataToBeSent[<index>] = null` race condition.

The tricky part is the race between `Dispose(bool disposing)` and `ReadOrRegisterCallback` in the `SerializedDataStream` type. Here is my analysis:
1. Both are guarded by `lock (_syncObject)`, so either `Dispose` runs before `ReadOrRegisterCallback` or `Dispose` runs after `ReadOrRegisterCallback`. In the latter case, calling `ReadOrRegisterCallback` should be safe. So, the former case is what we will focus on.
2. In the case that `Dispose` runs before `ReadOrRegisterCallback`, the call to `Read` in `ReadOrRegisterCallback` will fast return 0 because `_isDisposed == true`. Then, the `byte[]` object `result` will be returned from `ReadOrRegisterCallback`, which will be further returned to the caller of `PrioritySendDataCollection.ReadOrRegisterCallback`. But be noted that, the `result` in this case is byte array with all its elements to be `0`, because it was never used with in `Read`.

Now, the question is, does the caller of `PrioritySendDataCollection.ReadOrRegisterCallback` expect it would get a byte array with all its elements being `0`? I took a look at the call sites of `PrioritySendDataCollection.ReadOrRegisterCallback`, and the returned byte array is treated as legit payload and sent out in a few places. I guess that could be a problem.

So, I think we probably should check on `_isDisposed` in `SerializedDataStream.ReadOrRegisterCallback` as well, and fast return if the instance has been disposed.